src/draw/lv_draw_rect.h

```diff
@@ -33,12 +34,11 @@ typedef struct {
     lv_blend_mode_t blend_mode;
 
     /*Background*/
-    lv_color_t bg_color;
-    lv_color_t bg_grad_color;
-    uint8_t bg_main_color_stop;
-    uint8_t bg_grad_color_stop;
     lv_opa_t bg_opa;
-    lv_grad_dir_t bg_grad_dir : 3;
+    union {
```

Comment:
We needed to get rid of nameless unions due to some compiler issues. I don't remember exactly what was the problem. Maybe @embeddedt remembers.

Reply:
Anonymous unions are supported in C11, but not C99. In C99 they only work because of GNU extensions.

Reply:
So? What is the conclusion? Can I use it? 
I don't see a way to have both gradient and bg_color working at the same time without anonymous union.

Reply:
To be honest the only other solution is to use something a lot more ugly:
```cpp
struct {
    lv_color_t bg_color;
    uint8_t     _reserved[sizeof(lv_gradient_t) - sizeof(lv_color_t)];
    lv_gradient_t * grad; /**< Alias on bg_color above */
[...]
 
In constructor:
  grad = (lv_gradient_t*)&bg_color;
```

But, it's **SO** ugly compared to the union code.

Reply:
> Are we still restricted to 23 years old standard ?

Even with C99 it's challenging to support the wide variety of embedded compilers we have. It'd be great to move to C11, but I'm afraid we can't. :slightly_frowning_face: 

> What is the conclusion? Can I use it?

No. :slightly_frowning_face: 

> I don't see a way to have both gradient and bg_color working at the same time without anonymous union.

You suggestion seemed good to me:
```c
//In lv_draw_rect_dsc_t
lv_color_t bg_color;
lv_gradient_t grad; //stop_count is initialized to 0, and if it's 0 bg_color is used else grad
```

Reply:
This won't work well. Because if one decide to set the second stop later on, it's very likely he will forget to change stop_count and it'll fail. Also, it means he has to copy the color from bg_color to grad's stop. That's one more step to do so it's adding friction to the code, I don't like it.

*If only the field could be private so it would only be set via a method, but, eh, it's C, not C++.*
I also don't like the pointer in the structure like in the example above since all other field are not pointer, it's distracting.

And if I change the union not to be anonymous, it'll break API anyway:
```cpp
struct {
[...]
  union {
    lv_color_t color;
    lv_gradient_t grad;
  } bg;
[...]

// Ends up used like:
dsc.bg.color = ...
// or
dsc.bg.grad.stops[0].color = ...
```

Reply:
> This won't work well. Because if one decide to set the second stop later on, it's very likely he will forget to change stop_count and it'll fail. Also, it means he has to copy the color from bg_color to grad's stop. That's one more step to do so it's adding friction to the code, I don't like it.

Really not the best but I'd be ok with that as a compromise. In 99% of the cases people set only the `bg_color`. If rarely someone touches the gradients, he can take care a little.  If there were more stops the user needs to set the stop_count anyway.

Reply:
Ok, just commited a change. I'm using the `grad.dir` member as the test for the gradient presence. So setting gradients only require setting the 2 stops now and the direction (unlike previous code that required only setting the 2nd stop), not the stop count. It'll work for 99% of the cases, let's say it's ok.